index.js

```diff
@@ -1499,6 +1499,69 @@
     };
   }
 
+  //# validate :: Type -> a -> Either Object a
+  //.
+  //. Takes a type, and any value. Returns an `Array` of
+  //. `Right a` if the value is a member of the type;
+  //. `Left Object` for each property that is invalid.
+  //. The first index in an array of `Left`s is always named `$$`,
+  //. which refers to the entire value.
+  function validate(t) {
+    return function(x) {
+      var returnValue = [];
+      var $$Result    = t.validate ([]) (x);
+
+      returnValue.push (
+        $$Result.isLeft
+          ? Left ({
+              error: 'WrongValue',
+              type: t.name || t.type,
+              name: '$$',
+              value: x
+            })
+          : $$Result
+      );
+
+      if (x != null && t.keys.length > 0) {
+        var props = t.keys;
+
+        for (var i = 0, len = props.length; i < len; i += 1) {
+          var propName  = props[i];
+          var propValue = x[propName];
+          var type      = t.types[propName];
+          var result    = type.validate ([]) (propValue);
+
+          if (result.isLeft) {
+            if (propName in x) {
+              returnValue.push (
+                Left ({
+                  error: 'WrongValue',
+                  // TODO: figure what propPath really is
+                  type: result.value.propPath.length > 0
```

Comment:
@davidchambers this part I'm really not sure about. The idea is to get the name of the Type that is invalid for nested types. I haven't seen any `propPath` that didn't exactly contain `['$1']`but I'm sure there are cases.